From: Hannes Reinecke <hare@suse.de>
Date: Fri, 5 May 2017 12:16:29 +0200
Subject: fnic: Return 'DID_IMM_RETRY' if rport is not ready
Patch-Mainline: submitted linux-scsi 2017/05/22
References: bsc#1035920

According to the fcoe sources we should return 'DID_IMM_RETRY'
from queuecommand when the associated libfc rdata structures
are not present.

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 drivers/scsi/fnic/fnic_scsi.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/scsi/fnic/fnic_scsi.c b/drivers/scsi/fnic/fnic_scsi.c
index c28414b..ff49ce5 100644
--- a/drivers/scsi/fnic/fnic_scsi.c
+++ b/drivers/scsi/fnic/fnic_scsi.c
@@ -466,7 +466,17 @@ static int fnic_queuecommand_lck(struct scsi_cmnd *sc, void (*done)(struct scsi_
 	}
 
 	rp = rport->dd_data;
-	if (!rp || rp->rp_state != RPORT_ST_READY) {
+	if (!rp) {
+		/*
+		 * rport is transitioning from blocked/deleted to online
+		 */
+		FNIC_SCSI_DBG(KERN_DEBUG, fnic->lport->host,
+				"returning DID_IMM_RETRY for IO as rport is transitioning\n");
+		sc->result = DID_IMM_RETRY << 16;
+		done(sc);
+		return 0;
+	}
+	if (rp->rp_state != RPORT_ST_READY) {
 		FNIC_SCSI_DBG(KERN_DEBUG, fnic->lport->host,
 				"returning DID_NO_CONNECT for IO as rport is removed\n");
 		atomic64_inc(&fnic_stats->misc_stats.rport_not_ready);
-- 
1.8.5.6

